sky/spot/controller.py

```diff
@@ -88,12 +98,43 @@ def _run(self):
                 3. Any unexpected error happens during the `sky.launch`.
         Other exceptions may be raised depending on the backend.
         """
-        logger.info(f'Started monitoring spot job {self._job_id}, '
-                    f'name: {self._task_name!r}.')
-        spot_state.set_starting(self._job_id)
+        if task.run is None:
+            logger.info(f'Sub job {task_id} ({task.name}) is skipped '
+                        'for empty run.')
+            spot_state.set_started(self._job_id,
+                                   task_id,
+                                   start_time=time.time())
+            spot_state.set_succeeded(self._job_id,
+                                     task_id,
+                                     end_time=time.time())
+            return True
+        job_id_env_var = task.envs[constants.JOB_ID_ENV_VAR]
+        submitted_at = time.time()
+        if task_id == 0:
+            submitted_at = self._backend.run_timestamp
+        spot_state.set_submitted(
+            self._job_id,
+            task_id,
+            submitted_at,
+            resources_str=backend_utils.get_task_resources_str(task))
+        logger.info(f'Submitted spot job (sub job: {task_id}); '
+                    f'{constants.JOB_ID_ENV_VAR}: {job_id_env_var}')
+        logger.info(str(task))
+        assert task.name is not None, task
+        cluster_name = spot_utils.generate_spot_cluster_name(
+            task.name, self._job_id)
+        self._strategy_executor = recovery_strategy.StrategyExecutor.make(
+            cluster_name, self._backend, task, self._retry_until_up)
+
+        logger.info(f'Started monitoring spot job {self._job_id} (task: '
+                    f'{task_id}, name: {task.name!r}).')
+        spot_state.set_starting(self._job_id, task_id)
         job_submitted_at = self._strategy_executor.launch()
+        assert job_submitted_at is not None, job_submitted_at
```

Comment:
remnant? can it fail and return none?

Reply:
No, this is the start time of the job submitted to the remote spot VM. If it fails, it will raise corresponding error, such as `SpotJobReachedMaxRetryError` which will be handled by the caller in `self.run()`.

Reply:
Should we update launch()'s doc then?
```

        Returns: The job's submit timestamp, or None if failed.
```

Reply:
See the other comment.